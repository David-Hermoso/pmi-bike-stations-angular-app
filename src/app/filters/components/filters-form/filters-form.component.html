<form [formGroup]="filters">
  <div>
    <mat-form-field>
      <mat-label>Station name or address</mat-label>
      <input
        [formControl]="searchControl"
        type="text"
        matInput
        placeholder="Station name or address"/>
    </mat-form-field>
  </div>
  <div class="filters__form-actions">
    <button
      class="filters__reset-btn"
      mat-raised-button
      color="primary"
      (click)="resetFilters()">
      Reset filters
    </button>
    <button
      *ngIf="!isDesktopDevice"
      class="filters__apply-btn"
      mat-raised-button
      color="primary"
      (click)="applyFilters()">
      Apply filters
    </button>
  </div>
</form>
