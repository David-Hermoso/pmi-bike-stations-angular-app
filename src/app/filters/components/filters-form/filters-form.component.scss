.filters__distance-container {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.filters__distance-controls {
  display: flex;
}

.filters__distance-slider {
  width: 100%;
}

.filters__distance-field {
  width: 50px;
  margin-left: 1rem;
}

.filters__picture-field {
  margin-top: 1.25em;
}
.filters__form-actions {
  display: flex;
  justify-content: flex-end;
  margin-top: 1.25em;
}
.filters__apply-btn {
  margin-left: 1rem;
}
