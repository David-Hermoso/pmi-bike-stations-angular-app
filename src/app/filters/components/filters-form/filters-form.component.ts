import {Component, OnInit} from '@angular/core';
import {FormControl, FormGroup} from '@angular/forms';
import {Store} from '@ngrx/store';
import {AppState} from '../../../models/app.state';
import {filter} from '../../../state/actions/filters.actions';
import {DeviceDetectorService} from 'ngx-device-detector';
import {initialBikeStationsState} from '../../../state/reducers';
import {Subscription} from 'rxjs';

@Component({
  selector: 'app-filters-form',
  templateUrl: './filters-form.component.html',
  styleUrls: ['./filters-form.component.scss']
})
export class FiltersFormComponent {
  public kmToCityCenterControl: FormControl = new FormControl(initialBikeStationsState.filters.kmToCityCenter);
  public nameControl: FormControl = new FormControl(initialBikeStationsState.filters.name);
  public filters: FormGroup = new FormGroup({
    kmToCityCenter: this.kmToCityCenterControl,
    name: this.nameControl,
  });
  public isDesktopDevice = false;
  private filterFormValueChangesSubscription: Subscription | undefined;

  constructor(private store: Store<AppState>, private deviceService: DeviceDetectorService) {
    this.isDesktopDevice = deviceService.isDesktop();
    if (this.isDesktopDevice) {
      this.filterFormValueChangesSubscription = this.filters.valueChanges
        .subscribe(() => {
          this.applyFilters();
        });
    }
  }

  public resetFilters(): void {
    const initialFiltersState = initialBikeStationsState.filters;
    this.nameControl.patchValue(initialFiltersState.name);
    this.kmToCityCenterControl.patchValue(initialFiltersState.kmToCityCenter);

    if (!this.isDesktopDevice) {
      this.applyFilters();
    }
  }

  public applyFilters(): void {
    this.store.dispatch(filter({
        payload: {
          ...this.filters.value
        }
      })
    );
  }
}
