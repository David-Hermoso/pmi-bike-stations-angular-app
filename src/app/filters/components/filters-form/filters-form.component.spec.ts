import {of} from 'rxjs';
import {FiltersFormComponent} from './filters-form.component';

describe('FiltersFormComponent', () => {
  describe('mobile', () => {
    const deviceDetectorMock = {
      isDesktop: () => false
    };
    const mockStore = {
      dispatch: jest.fn(),
      pipe: jest.fn(() => of([
        {
          img: null,
          id: '09',
          name: 'F. MANUEL HERREROS',
          lng: 2.664034367,
          lat: 39.57458439,
          fullAddress: 'Francesc Manuel de los Herreros, 32 07005 Palma, Illes Balears',
          street: 'Francesc Manuel de los Herreros',
          streetNumber: 32,
          cp: '07005',
          town: 'Palma',
          region: 'Illes Balears'
        },
        {
          img: null,
          id: '41',
          name: 'FÁBRICA',
          lng: 2.637383938,
          lat: 39.57274848,
          fullAddress: 'Comte de Barcelona, 25, 07013 Palma, Illes Balears',
          street: 'Comte de Barcelona',
          streetNumber: 25,
          cp: '07013',
          town: 'Palma',
          region: 'Illes Balears'
        }
      ]))
    };
    const component = new FiltersFormComponent(mockStore as any, deviceDetectorMock as any);

    it('isDesktopDevice should be false', () => {
      expect(component.isDesktopDevice).toBeFalsy();
    });
    it('filters should not be dispatched', done => {
      component.filters.valueChanges.subscribe(() => {
        expect(mockStore.dispatch).not.toHaveBeenCalled();
        done();
      });

      component.filters.patchValue({
        name: 'sant',
      });
    });

    it('resetFilters: filter action should be dispatched', () => {
      component.resetFilters();

      expect(mockStore.dispatch).toHaveBeenCalled();
    });

    it('applyFilters: filter action should be dispatched', () => {
      component.resetFilters();

      expect(mockStore.dispatch).toHaveBeenCalled();
    });
  });
  describe('desktop', () => {
    const deviceDetectorMock = {
      isDesktop: () => true
    };
    const mockStore = {
      dispatch: jest.fn(),
      pipe: jest.fn(() => of([
        {
          img: null,
          id: '47',
          name: 'PL PORTA S CATALINA',
          lng: 2.641439438,
          lat: 39.57113582,
          fullAddress: 'Porta de Santa Catalina, 17, 07012 Palma, Illes Balears',
          street: 'Jaume III',
          streetNumber: 17,
          cp: '07012',
          town: 'Palma',
          region: 'Illes Balears'
        },
        {
          img: null,
          id: '49',
          name: 'PL DE LA REINA',
          lng: 2.645848989,
          lat: 39.56797655,
          fullAddress: 'Antoni Maura, 10, 07012 Palma, Illes Balears',
          street: 'Antoni Maura',
          streetNumber: 10,
          cp: '07012',
          town: 'Palma',
          region: 'Illes Balears'
        }
      ]))
    };
    const component = new FiltersFormComponent(mockStore as any, deviceDetectorMock as any);

    it('isDesktopDevice should be true', () => {
      expect(component.isDesktopDevice).toBeTruthy();
    });
    it('filters should be dispatched', done => {
      component.filters.valueChanges.subscribe(() => {
        expect(mockStore.dispatch).toHaveBeenCalled();
        done();
      });

      component.filters.patchValue({
        name: 'sant',
      });
    });
  });
  it('resetFilters: form control should have initial state values', () => {
    const deviceDetectorMock = {
      isDesktop: () => true
    };
    const mockStore = {
      dispatch: jest.fn(),
      pipe: jest.fn(() => of([
        {
          img: null,
          id: '49',
          name: 'PL DE LA REINA',
          lng: 2.645848989,
          lat: 39.56797655,
          fullAddress: 'Antoni Maura, 10, 07012 Palma, Illes Balears',
          street: 'Antoni Maura',
          streetNumber: 10,
          cp: '07012',
          town: 'Palma',
          region: 'Illes Balears'
        }
      ]))
    };
    const component = new FiltersFormComponent(mockStore as any, deviceDetectorMock as any);

    component.resetFilters();

    expect(component.filters.value).toMatchSnapshot();
  });
});
