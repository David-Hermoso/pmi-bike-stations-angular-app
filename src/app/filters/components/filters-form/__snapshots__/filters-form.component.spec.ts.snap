// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiltersFormComponent resetFilters: form control should have initial state values 1`] = `
Object {
  "kmToCityCenter": 20,
  "name": "",
}
`;
